Implement script component and processor

iamuchejude opened the pull request:
#### Description of what you did:

Implement a script processor to execute script tags from embeds.

#### My PR is a:

- 🔝 Improvement

#### Is the PR ready to be merged?

- ✅ Yes!

#### Feature Request

https://community.frontity.org/t/execute-script-tags-found-in-content/1060

#### Docs Changes
https://app.gitbook.com/@frontity-tech/s/frontity/~/drafts/-LzGMLsfZI_8PDDtUgAv/api-reference-1/frontity-components/@drafts

----

Fixes #284.
Fixes #283.

iamuchejude committed:
feat(html2react): implement script processor

luisherranz commented on packages/components/script.tsx:
```diff
@@ -0,0 +1,21 @@
+
+
+    window.document.body.appendChild(script)
+  }
```
I think this needs to be added inside a `useEffect` with no dependencies (`[]`) so we don't add the same script over and over on each React re-rerender.

luisherranz reviewed (changes requested):
Looks great to me so far @iamuchejude :)

We use [`changesets`](https://github.com/atlassian/changesets) for versioning. You need to [add a changeset](https://github.com/atlassian/changesets/blob/master/docs/adding-a-changeset.md) to this PR. Take a look at the changeset docs and if you have any doubt let us know.

I've also left a comment about a change I think we should make.

Apart from that, maybe we can have a quick call tomorrow to see how to add e2e tests for this.

iamuchejude committed:
fix(html2react): script component re-render

luisherranz commented on packages/components/script.tsx:
```diff
@@ -0,0 +1,34 @@
+const generateId = () => Math.trunc(Math.random() * 9999999).toString();
+
+const Script: ScriptType = ({ state, ...props }) => {
+  const scriptId = useRef(generateId());
```
Instead of `generateId` which will break hydration I would try to keep the reference to the `script`. Something like this:

```js
const Script: ScriptType = ({ state, ...props }) => {
  let script = null;
  const { src } = props;

  useEffect(() => {
    if (state.frontity.rendering === "csr") {
      script = window.document.createElement("script");
      script.src = src;
      script.async = true;
      script.charset = "utf-8";

      window.document.body.appendChild(script);
    }

    return () => {
      if (script) window.document.body.removeChild(script);
    };
  }, []);

  return <script async src={src} charSet="utf-8" />;
};
```

luisherranz commented on packages/html2react/processors/image.tsx:
```diff
@@ -23,6 +23,8 @@ const image: Processor = {
 export default image;
 
 const ContentImage: React.FC<any> = props => {
+  console.log(props);
```
Careful with the console logs.

luisherranz commented on packages/mars-theme/src/index.js:
```diff
@@ -44,7 +45,7 @@ const marsTheme = {
        * inside the content HTML. You can add your own processors too
        */
-      processors: [image]
+      processors: [image, script]
```
Instead of adding the processor on each theme, I think we can add it by default in `html2react`:
https://github.com/frontity/frontity/blob/dev/packages/html2react/src/index.ts#L11

because we are "fixing something we broke". It should be fixed by default.

luisherranz commented on packages/html2react/processors/script.tsx:
```diff
@@ -0,0 +1,20 @@
+import Script from '@frontity/components/script';
+
+const script: Processor = {
+  test: node => node.type === "element" && node.component === "script",
```
We need to add a name (I guess `"script"` it's fine) and a priority, which I think it should be at least 20 so other people can overwrite this.

iamuchejude committed:
chore(component): refactor script component

iamuchejude committed:
ch(changeset): add changeset

iamuchejude committed:
chore(html2react): make script processor built-in

luisherranz commented:
Right now it looks like the best idea is to `return null` in the `<Script>` component because if not it will be executed/loaded twice in SSR. We need to figure out how to make it work with SSR when Frontity doesn't load.

iamuchejude committed:
ch(e2e): write e2e test for script component
ch(component): fix script duplicacy on pages
ft(proccessor): add support for rendering on inline script

iamuchejude committed:
fix(e2e): fix e2e test for script component

luisherranz commented on examples/mars-theme-example/frontity.settings.js:
```diff
@@ -32,7 +32,7 @@ export default {
       state: {
         source: {
-          api: "https://test.frontity.io/wp-json"
+          api: "http://frontity.local/wp-json"
```
You need to revert this change.

luisherranz commented on packages/components/script.tsx:
```diff
@@ -0,0 +1,35 @@
+  const { src, code, dataId } = props;
+
+  useEffect(() => {
+    if (state.frontity.rendering === "csr") {
```
I think there's not much sense to use `state.frontity.rendering` now that we are returning null.

luisherranz commented on packages/components/script.tsx:
```diff
@@ -0,0 +1,35 @@
+      script = window.document.createElement("script");
+
+      if (src) script.src = src;
+      if (code) eval(code);
```
I think we don't need to add a script tag if what we are going to do is to use `eval` for the code.

luisherranz commented on packages/components/script.tsx:
```diff
@@ -0,0 +1,35 @@
+
+      if (src) script.src = src;
+      if (code) eval(code);
+      if (dataId) script.id = dataId;
```
Is there any reason to use `dataId` instead of `id` directly?

luisherranz commented on packages/components/script.tsx:
```diff
@@ -0,0 +1,35 @@
+      if (dataId) script.id = dataId;
+
+      script.async = true;
+      script.charset = "utf-8";
```
We can remove `charset`. It looks like it's been deprecated: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/script

luisherranz reviewed (changes requested):
Looks great @iamuchejude :)

Last things we can explore:
- Do e2e tests for the processor.
- Pass down other `script` props, like `type`.

iamuchejude committed:
chore(script): refactor script component

iamuchejude committed:
Replace source api URL

luisherranz commented:
Let's only apply the processor if type is missing or it's javascript (text/javascript, application/javascript...).

iamuchejude committed:
chore(html2react): update test for script processor

luisherranz committed:
Html2React: console.warn the errors in processor test

There was no way to know if a processor test was failing.

luisherranz committed:
Html2React: First unit tests for script processor

luisherranz committed:
Html2React: several fixes to the script processor

- Add a name to the processor.
- Fix error when the node doesn't have a parent.
- Don't set the code prop if there are no children.
- Pass down the correct content of children in the code prop.
- If the code prop is present, don't pass down children.

luisherranz committed:
Switch to regular snapshots & add invalid type test

iamuchejude committed:
Html2React: several updates to the script processor

- Don't pass data-src props to Script component
- Write unit test to check if src is set when data-src is provided

Others: Add .vscode folder to the .gitignore file

luisherranz reviewed (approved):
Great work @iamuchejude, ready for merge :)

Is the code finished?

luisherranz committed:
Merge branch 'dev' into script-processor

luisherranz commented:
I've merged `dev` to solve a problem with the cypress library not working in Ubuntu 18.04.

iamuchejude commented:
@luisherranz The PR is ready to be merged. :)

luisherranz commented:
Awesome. Thanks Jude!

If we merge this PR it will close the related issues and the "update docs" tasks are not finished yet so we need to figure out how to wait and review the documentation. I'm going to wait until I talk with @SantosGuillamot about this.

iamuchejude commented:
Thanks, Luis! 🙌

I'd have a meeting with @SantosGuillamot on Monday to talk about documentation. I think that's all that is left for the issues to be resolved and closed.

luisherranz committed:
Move testing library to devDependencies

luisherranz committed:
Script: add e2e for load of the requested script

luisherranz committed:
Script: add unit test to check valid types

luisherranz reviewed (changes requested):
Great work @iamuchejude.

- I've added an e2e test to check that the script is actually being loaded.
- I've moved the react testing library dependency to `devDependencies`.
- I've added a unit test to check if the valid types are being picked up fine, but it fails. Can you check it out?

I'll review the docs next 👍

iamuchejude commented:
Okay @luisherranz

luisherranz reviewed (changes requested):
Docs are great, thanks. Could you please add a couple of examples of usage of the `<Script>` component? One with `src` and another with the `code` prop. Thanks! :)

iamuchejude committed:
HTML2React: Fix failing unit tests

iamuchejude committed:
Merge branch 'dev' into script-processor

luisherranz committed:
Add TypeScript props to processors

luisherranz committed:
Remove package-lock and lerna log

luisherranz committed:
Merge branch 'dev' into script-processor

luisherranz reviewed (approved):
Thanks @iamuchejude!

I've added TypeScript props to the Html2react processor but other than that it's fine, thanks!

Be careful with committing the lerna logs and package-lock files (we must add those to the .gitignore file).

luisherranz commented:
One test is failing. I'll fix it later.

luisherranz committed:
Add script processor by default

This was removed accidentally on the last merge with dev.

luisherranz commented:
Fixed. `script` was missing because I accidentally removed it when I merged with the latest `dev` branch.
